clickhouse-admin/src/context.rs

```diff
@@ -3,84 +3,555 @@
+        if current > incoming_generation {
+            return Err(HttpError::for_client_error(
+                Some(String::from("Conflict")),
+                StatusCode::CONFLICT,
+                format!(
+                    "current generation '{}' is greater than incoming generation '{}'",
+                    current,
+                    incoming_generation,
+                )
+            ));
+        }
+    };
+
+    let output = match node_settings {
+        NodeSettings::Replica { ref settings, .. } => {
+            GenerateConfigResult::Replica(
+                clickward.generate_server_config(settings)?,
+            )
+        }
+        NodeSettings::Keeper { ref settings, .. } => {
+            GenerateConfigResult::Keeper(
+                clickward.generate_keeper_config(settings)?,
+            )
+        }
+    };
+
+    // We want to update the generation number only if the config file has been
+    // generated successfully.
+    generation_tx.send(Some(incoming_generation)).map_err(|e| {
+        HttpError::for_internal_error(format!("failure to send request: {e}"))
+    })?;
+
+    // Once we have generated the client we can safely enable the clickhouse_server service
+    Svcadm::enable_service(node_settings.fmri())?;
+
+    Ok(output)
+}
+
+pub async fn init_db(
+    clickhouse_address: SocketAddrV6,
+    log: Logger,
+    replicated: bool,
+) -> Result<(), HttpError> {
+    // We initialise the Oximeter client here instead of keeping it as part of the context as
+    // this client is not cloneable, copyable by design.
+    let client = OximeterClient::new(clickhouse_address.into(), &log);
+
+    // Initialize the database only if it was not previously initialized.
+    // TODO: Migrate schema to newer version without wiping data.
+    let version = client.read_latest_version().await.map_err(|e| {
+        HttpError::for_internal_error(format!(
+            "can't read ClickHouse version: {e}",
+        ))
+    })?;
+    if version == 0 {
+        info!(
+            log,
+            "initializing oximeter database to version {OXIMETER_VERSION}"
+        );
+        client
+            .initialize_db_with_version(replicated, OXIMETER_VERSION)
+            .await
+            .map_err(|e| {
+            HttpError::for_internal_error(format!(
+                "can't initialize oximeter database \
+                     to version {OXIMETER_VERSION}: {e}",
+            ))
+        })?;
+    } else {
+        info!(
+            log,
+            "skipping initialization of oximeter database at version {version}"
+        );
+    }
+
+    Ok(())
+}
+
+fn read_generation_from_file(path: Utf8PathBuf) -> Result<Option<Generation>> {
+    // When the configuration file does not exist yet, this means it's a new server.
+    // It won't have a running clickhouse server and no generation number yet.
+    if !path.exists() {
+        return Ok(None);
+    }
+
+    let file = File::open(&path)?;
```

Comment:
Can we add context to this error? Something like

```suggestion
    let file = File::open(&path).with_context(|| format!("failed to open {path}"))?;
```